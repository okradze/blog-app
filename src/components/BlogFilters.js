import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { setTitleFilter, searchByAuthor, searchByTitle } from '../actions/filters';

export const BlogFilters = (props) => {
    const onTitleChange = (e) => {
        props.setTitleFilter(e.target.value);
    };

    const onSortChange = (e) => {
        if(e.target.value === 'title') {
            props.searchByTitle();
        } else if (e.target.value === 'author') {
            props.searchByAuthor();
        }
    };

    useEffect(() => {
        return () => {
            props.setTitleFilter('');
        }
    }, [])

    return (
        <div className="flex items-center">
            <input type="text" className="py-2 px-4 mr-4 rounded-sm" placeholder="Search by" value={props.filters.title} onChange={onTitleChange} />
            <select className="py-2 px-4 rounded-sm" value={props.filters.searchBy} onChange={onSortChange}>
                <option value="title">Title</option>
                <option value="author">Author</option>
            </select>
        </div>
    );
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTitleFilter: (title) => dispatch(setTitleFilter(title)),
    searchByAuthor: () => dispatch(searchByAuthor()),
    searchByTitle: () => dispatch(searchByTitle())
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogFilters);