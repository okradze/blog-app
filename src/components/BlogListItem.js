import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import 'intersection-observer';

export const BlogListItem = ({ title, author, id, to, createdAt, url, photoURL }) => {
	const [ visibleCaption, setVisibleCaption ] = useState(true);

	return (
		<article className="blog-item">
			<div className="blog-info">
				<img src={photoURL} className="blog-info__photo" alt={author} />
				<div className="blog-info__content">
					<p className="blog-info__author">{author}</p>
					<time dateTime={new Date(createdAt)} className="blog-info__date">
						{moment(createdAt).format('MMM Do, YYYY')}
					</time>
				</div>
			</div>

			<figure className="blog-item__thumbnail">
				{visibleCaption && <div className="blog-item__thumbnail-no" />}
				<img onLoad={() => setVisibleCaption()} src={url} alt={title} className="blog-item__thumbnail-img" />
			</figure>

			<div className="blog-item__content">
				<h2 className="blog-item__title">{title}</h2>
				{to === 'edit' && (
					<Link className="read-more" to={`${to}/${id}`}>
						Edit Blog
					</Link>
				)}
				<Link className="read-more" to={`read/${id}`}>
					Read more...
				</Link>
			</div>
		</article>
	);
};
export default BlogListItem;
