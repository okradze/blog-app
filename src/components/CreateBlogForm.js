import React, { useState } from 'react';
import { connect } from 'react-redux';
import xss from 'xss';
import Quill from './Quill';
import loader from '../images/loader.gif';

export const CreateBlogForm = ({
	onSubmit,
	title,
	body,
	createdAt,
	submitDisabled,
	submitLoading,
	isEdit,
	removeDisabled,
	removeLoading,
	onRemove,
	auth,
}) => {
	const [ blogTitle, setTitle ] = useState(title || '');
	const [ blogBody, setBody ] = useState(body || '');
	const [ blogCreatedAt ] = useState(createdAt || new Date().getTime());
	const [ file, setFile ] = useState(null);
	const [ error, setError ] = useState('');

	const onTitleChange = e => {
		setTitle(e.target.value);
	};
	const onBodyChange = value => {
		setBody(value);
	};
	const onSave = e => {
		e.preventDefault();

		const data = {
			title: xss(blogTitle),
			body: xss(blogBody),
			createdAt: blogCreatedAt,
			author: xss(auth.displayName),
			photoURL: xss(auth.photoURL),
			file,
		};

		if (isEdit) {
			if (!blogTitle || blogBody.length < 200) {
				setError('Please provide blog title and make sure body contains more than 200 characters');
			} else {
				setError('');
				onSubmit(data, auth.uid);
			}
		} else if (!isEdit) {
			if (!blogTitle || blogBody.length < 200 || !file) {
				setError('Please provide blog title, thumbnail and make sure body contains more than 200 characters');
			} else {
				setError('');
				onSubmit(data, auth.uid);
			}
		}
	};

	const onFileChange = e => {
		if (e.target.files[0]) {
			setFile(e.target.files[0]);
		}
	};

	return (
		<form className="create-form" onSubmit={onSave}>
			{error && (
				<div className="alert create-form__alert" role="alert">
					<p>{error}</p>
				</div>
			)}
			<input
				type="text"
				className="input create-form__input"
				placeholder="Blog title"
				value={blogTitle}
				onChange={onTitleChange}
			/>

			{!isEdit && (
				<input type="file" className="input file-upload" onChange={onFileChange} multiple accept="image/*" />
			)}

			<Quill body={blogBody} setBody={onBodyChange} />
			<button onClick={onSave} type="button" className="button create-form__button" disabled={submitDisabled}>
				Save Blog
			</button>
			{submitLoading && <img className="small-loader form-loader" src={loader} alt="loader" />}
			{isEdit && (
				<React.Fragment>
					<button type="button" className="button button--grey" onClick={onRemove} disabled={removeDisabled}>
						Remove
					</button>
					{removeLoading && <img className="small-loader form-loader" src={loader} alt="loader" />}
				</React.Fragment>
			)}
		</form>
	);
};

const mapStateToProps = state => ({
	auth: state.auth,
});

export default connect(mapStateToProps)(CreateBlogForm);
