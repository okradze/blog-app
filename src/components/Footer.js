import React from 'react';
import { Link } from 'react-router-dom';

export const Footer = () => (
	<footer className="footer">
		<div className="container footer__content">
			<nav className="footer__nav">
				<Link className="footer__link" to="/dashboard">
					Dashboard
				</Link>
				<Link className="footer__link" to="/create">
					Create
				</Link>
				<Link className="footer__link" to="/read">
					Blogs
				</Link>
			</nav>
			<p className="footer__copy">Copyright &copy; 2019 Mirian Okradze</p>
		</div>
	</footer>
);

export default Footer;
