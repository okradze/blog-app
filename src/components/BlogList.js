import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import 'intersection-observer';
import { fetchBlogs, firstFetchBlogs } from '../services/fetch';
import BlogListItem from './BlogListItem';
import loader from '../images/loader.gif';
import selectBlogs from '../selectors/blogs';

export const BlogList = ({ uid, filters }) => {
	const [ blogs, setBlogs ] = useState([]);
	const [ loading, setLoading ] = useState(true);
	const [ fetching, setFetching ] = useState(false);

	const scrollElement = React.useRef();

	const paginateBlogs = () => {
		setFetching(true);
		fetchBlogs(uid)
			.then(newBlogs => {
				const data = JSON.parse(sessionStorage.getItem('blogs'));
				setBlogs([ ...data, ...newBlogs ]);
				sessionStorage.setItem('blogs', JSON.stringify([ ...data, ...newBlogs ]));
				setFetching(false);
			})
			.catch(() => {
				setFetching(false);
			});
	};

	useEffect(() => {
		let didCancel = false;
		firstFetchBlogs(uid).then(newBlogs => {
			if (!didCancel) {
				setBlogs(newBlogs);
				sessionStorage.setItem('blogs', JSON.stringify(newBlogs));
				setLoading(false);

				const io = new IntersectionObserver(entries => {
					const ratio = entries[0].intersectionRatio;
					if (ratio > 0) {
						setFetching(true);
						paginateBlogs();
					}
				});
				io.observe(scrollElement.current);
			}
		});
		return () => {
			didCancel = true;
		};
	}, []);

	const filteredBlogs = selectBlogs(blogs, filters);

	return (
		<React.Fragment>
			{loading ? (
				<img className="center small-loader" src={loader} alt="Loader" />
			) : (
				<div className="blog-list">
					{filteredBlogs.map(blog => <BlogListItem to="edit" key={blog.id} {...blog} />)}
					{fetching && <img className="small-loader" src={loader} alt="Loader" />}
					<div style={{ minHeight: '1px', minWidth: '1px' }} ref={scrollElement} />
				</div>
			)}
		</React.Fragment>
	);
};

const mapStateToProps = state => ({
	uid: state.auth.uid,
	filters: state.filters,
});

export default connect(mapStateToProps)(BlogList);
