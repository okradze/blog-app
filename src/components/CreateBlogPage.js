import React, { useState } from 'react';
import CreateBlogForm from './CreateBlogForm';
import { startCreateBlog } from '../actions/blogs';

export const CreateBlogPage = ({ history }) => {
	const [ submitLoading, setSubmitLoading ] = useState(false);
	const [ submitDisabled, setSubmitDisabled ] = useState(false);

	const onSubmit = (blog, uid) => {
		setSubmitLoading(true);
		setSubmitDisabled(true);
		startCreateBlog(blog, uid).then(() => {
			history.push('/dashboard');
		});
	};
	return (
		<React.Fragment>
			<div className="container mb-auto">
				<CreateBlogForm submitDisabled={submitDisabled} submitLoading={submitLoading} onSubmit={onSubmit} />
			</div>
		</React.Fragment>
	);
};

export default CreateBlogPage;
