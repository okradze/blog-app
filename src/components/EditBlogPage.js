import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import NotFoundPage from './NotFoundPage';
import CreateBlogForm from './CreateBlogForm';
import { startRemoveBlog, startEditBlog } from '../actions/blogs';
import loader from '../images/loader.gif';
import db from '../firebase/firebase';

const fetchBlog = async (uid, id) => {
	const data = await db.collection('users').doc(uid).collection('blogs').doc(id).get();
	return {
		id: data.id,
		...data.data(),
	};
};

export const EditBlogPage = ({ uid, match, history }) => {
	const [ removeLoading, setRemoveLoading ] = useState(false);
	const [ submitLoading, setSubmitLoading ] = useState(false);
	const [ submitDisabled, setSubmitDisabled ] = useState(false);
	const [ removeDisabled, setRemoveDisabled ] = useState(false);
	const [ loading, setLoading ] = useState(true);
	const [ blog, setBlog ] = useState(null);

	useEffect(() => {
		let didCancel = false;
		fetchBlog(uid, match.params.id).then(newBlog => {
			if (!didCancel) {
				setBlog(newBlog);
				setLoading(false);
			}
		});
		return () => {
			didCancel = true;
		};
	}, []);

	const onRemove = () => {
		setRemoveLoading(true);
		setRemoveDisabled(true);
		setSubmitDisabled(true);

		startRemoveBlog(uid, blog.id).then(() => {
			history.push('/dashboard');
		});
	};
	const onSubmit = updates => {
		setSubmitDisabled(true);
		setSubmitLoading(true);
		setRemoveDisabled(true);

		startEditBlog(uid, blog.id, updates).then(() => {
			history.push('/dashboard');
		});
	};

	return loading ? (
		<img className="center small-loader" src={loader} alt="Loader" />
	) : (
		<React.Fragment>
			{blog ? (
				<div className="container mb-auto">
					<CreateBlogForm
						submitLoading={submitLoading}
						submitDisabled={submitDisabled}
						{...blog}
						onSubmit={onSubmit}
						isEdit={true}
						removeLoading={removeLoading}
						removeDisabled={removeDisabled}
						onRemove={onRemove}
					/>
				</div>
			) : (
				<NotFoundPage />
			)}
		</React.Fragment>
	);
};

const mapStateToProps = state => ({
	uid: state.auth.uid,
});

export default connect(mapStateToProps)(EditBlogPage);
