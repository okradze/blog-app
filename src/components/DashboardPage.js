import React from 'react';
import 'babel-polyfill';
import BlogFilters from './BlogFilters';
import BlogList from './BlogList';

const DashboardPage = () => (
	<div className="dashboard-page container">
		<BlogFilters />
		<BlogList />
	</div>
);

export default DashboardPage;
