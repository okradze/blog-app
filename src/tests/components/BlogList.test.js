import React from 'react';
import 'babel-polyfill';
import { shallow } from 'enzyme';
import { BlogList } from '../../components/BlogList';

jest.mock('../../services/fetch');

let uid, filters;

beforeEach(() => {
	uid = 'abc123';
	filters = {
		title: '',
		searchBy: 'title',
	};
});

test('should render component when loading', () => {
	const wrapper = shallow(<BlogList uid={uid} filters={filters} />);
	expect(wrapper).toMatchSnapshot();
});

test('should render component when loading', done => {
	const wrapper = shallow(<BlogList uid={uid} filters={filters} />);
	setTimeout(() => {
		wrapper.update();
		expect(wrapper).toMatchSnapshot();
		done();
	}, 0);
});
