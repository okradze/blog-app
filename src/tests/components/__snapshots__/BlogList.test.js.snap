// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component when loading 1`] = `
<Fragment>
  <img
    alt="Loader"
    className="center small-loader"
    src={Object {}}
  />
</Fragment>
`;

exports[`should render component when loading 2`] = `
<Fragment>
  <img
    alt="Loader"
    className="center small-loader"
    src={Object {}}
  />
</Fragment>
`;
